documentation/specs/question.md

```diff
@@ -0,0 +1,84 @@
+
+```C#
+if (FailIfNotIncremental)
+{
+  TaskLoggingHelper.LogErrorWithCodeFromResources("ToolTask.NotUpToDate");
+  return false;
+}
+else
+{
+  TaskLoggingHelper.LogMessageWithCodeFromResources("ToolTask.NotUpToDate");
+}
+```
+
+
+### `ToolTask`
+If inheriting from ToolTask, your custom task can override `SkipTaskExecution()`.  When it returns `false`, TookTask will exit with error.
+
+## Shipping Tasks
+When `/question` switch is used, it will modify the shipping task with these behavior.  Note: this is still experimental and can change.
+
+`Exec`
+Doesn't have an up-to-date check.  It will always run.  Use Target Inputs and Outputs to skip the exec task.
+
+`Touch`
+Warns when a file is touched.  It is unclear if the file touched will participate in the build as it is a common practice to touch a file to signal external tool to run.  Use Target Inputs and Outputs to skip this task.
+
+`Copy`
+Errors out when any copy action occurs.
+
+`WriteLinesToFile`
+Error when WriteOnlyWhenDifferent is true.  This task could be used to append to a log file that isn't participating in the build itself.
+
+`Delete`
+Warn that a file still exists and is to be deleted.  It is unclear if the file is part of the build or it is deleting an accessary file outside of the build.
+
+`Move`
+No warning or errors.  This task is effectively a no-op as files could not exist anymore.
+
+`DownloadFile`
+Error if SkipUnchangedFiles is true.
+
+`GenerateResource` 
+Error if any files would be generated.
+
+`MakeDir`
+Error if folder doesn't exist.
+
+`RemoveDir`
+Error if folder still exist.
+
+`Unzip`
+Error if SkipUnchangedFiles is true.
+
+`ZipDirectory`
+Error if the destination zip file doesn't exists.
+
+## Common Up-To-Date Error
+- **Typographical error**. Spelling or incorrect path.  Check if the target inputs and outputs real files.
+- **Casing**. While MSBuild on Windows is case insenstive, some command line comparision and hashing tool are case senstive.  Paths and Project Refernces are a common source of casing issues.
+- **Cross Product**. Inputs and Outputs are sometimes used for Cross Product. Try to move all to `Outputs` parameter or `Returns` parameter instead of Inputs.
+- **Exec Task** are not Skipable, thus they should be wrapped with Target Inputs and Outputs or other systems.  For backwards compatibility, Question will not issue an error.
+- **FileWritten**.  The common clean system will remove files that aren't in the FileWritten itemgroup.  Sometimes task output won't be add to FileWritten itemgroup.
+- **Build, Build, Build**.  Sometimes, a 2nd build will expose up to date issue.  Verify building 3 times. For example, 1st Build, Question 2nd build, Run 2nd Build, and Question 3rd build.
```

Comment:
The numbers are confusing here - should it be ``Build 3 times, after each run, run the build with `/question` switch` to reveal possible incrementality issues.`` ?